File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    // this is my own thought process
    void helper(vector<string> &ans, string str, int opens, int closes) {
        if (opens == 0 && closes == 0) {
            ans.push_back(str);
            return;
        }
        if (opens > 0)
            helper(ans, str+"(", opens - 1, closes);
        if (opens < closes)
            helper(ans, str+")", opens, closes - 1);
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        helper(ans, "", n, n);
        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.cs
public class Solution {
    private static void myHelper(IList<string> res, string str, int opens, int closes) {
        if (opens == 0 && closes == 0) {
            res.Add(str);
            return;
        }
        if (opens > 0)
            myHelper(res, str + "(", opens - 1, closes);
        if (opens < closes)
            myHelper(res, str + ")", opens, closes - 1);
    }
    
    public IList<string> GenerateParenthesis(int n) {
        IList<string> res = new List<string>();
        myHelper(res, "", n, n);
        return res;
    }
}
